File: app/src/main/java/com/example/phone_stats_collector/MainActivity.java
package com.example.phone_stats_collector;

import android.Manifest;
import android.app.AlertDialog;
import android.content.Intent;
import android.content.pm.PackageManager;
import android.os.Build;
import android.os.Bundle;
import android.os.Handler;
import android.os.Looper;
import android.provider.Settings;
import android.text.method.ScrollingMovementMethod;
import android.util.Log;
import android.widget.Button;
import android.widget.ScrollView;
import android.widget.TextView;
import android.widget.Toast;

import androidx.appcompat.app.AppCompatActivity;
import androidx.core.app.ActivityCompat;
import androidx.core.content.ContextCompat;

import java.util.Locale;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;

public class MainActivity extends AppCompatActivity {
    private static final String TAG = "MainActivity";
    private static final int REQUEST_BLUETOOTH = 1001;

    private TextView tvInfo;
    private Button btnRefresh;
    private Button btnUpdateCheck;
    private ScrollView scrollView;

    private final ExecutorService executor = Executors.newSingleThreadExecutor();
    private final Handler mainHandler = new Handler(Looper.getMainLooper());

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        tvInfo = findViewById(R.id.tv_info);
        btnRefresh = findViewById(R.id.btn_refresh);
        btnUpdateCheck = findViewById(R.id.btn_update);
        scrollView = findViewById(R.id.scroll_view);

        tvInfo.setMovementMethod(new ScrollingMovementMethod());
        tvInfo.setTextIsSelectable(true);

        btnRefresh.setOnClickListener(v -> collectStats());
        btnUpdateCheck.setOnClickListener(v -> checkForUpdate());

        // 첫 실행 시 권한 체크
        checkPermissions();
        collectStats();
    }

    /** 필요 권한들을 한번에 체크 + 요청 */
    private void checkPermissions() {
        // 1. UsageStats 권한 (special permission — 설정으로 직접 이동)
        if (!hasUsageStatsPermission()) {
            showUsageStatsDialog();
            return;
        }

        // 2. Bluetooth 권한 (Android 12+)
        if (Build.VERSION.SDK_INT >= 31) {
            if (ContextCompat.checkSelfPermission(this, Manifest.permission.BLUETOOTH_CONNECT)
                    != PackageManager.PERMISSION_GRANTED) {
                ActivityCompat.requestPermissions(this,
                        new String[]{Manifest.permission.BLUETOOTH_CONNECT},
                        REQUEST_BLUETOOTH);
            }
        }
    }

    /** UsageStats 권한 보유 여부 */
    private boolean hasUsageStatsPermission() {
        if (Build.VERSION.SDK_INT < Build.VERSION_CODES.LOLLIPOP) return true;
        try {
            android.app.usage.UsageStatsManager usm =
                    (android.app.usage.UsageStatsManager) getSystemService(USAGE_STATS_SERVICE);
            if (usm == null) return false;
            long end = System.currentTimeMillis();
            long start = end - 1000;
            return usm.queryUsageStats(
                    android.app.usage.UsageStatsManager.INTERVAL_DAILY,
                    start, end) != null;
        } catch (Exception e) {
            return false;
        }
    }

    /** UsageStats 권한 요청 다이얼로그 */
    private void showUsageStatsDialog() {
        new AlertDialog.Builder(this)
                .setTitle("📊 사용량 접근 권한 필요")
                .setMessage("최근 사용 앱 목록을 수집하려면 '사용량 접근' 권한이 필요합니다.\n\n" +
                        "설정 → 사용량 접근 → BGMonitor 활성화 해주세요.")
                .setPositiveButton("설정으로 이동", (d, w) -> {
                    Intent intent = new Intent(Settings.ACTION_USAGE_ACCESS_SETTINGS);
                    startActivity(intent);
                    Toast.makeText(this, "권한 활성화 후 새로고침 해주세요", Toast.LENGTH_LONG).show();
                })
                .setNegativeButton("다음에", null)
                .setCancelable(false)
                .show();
    }

    private void collectStats() {
        btnRefresh.setEnabled(false);
        btnRefresh.setText("수집 중...");

        executor.execute(() -> {
            StringBuilder sb = new StringBuilder();

            sb.append("━━━ 📱 기기 정보 ━━━\n");
            sb.append("  기종: ").append(Build.MODEL).append("\n");
            sb.append("  제조사: ").append(Build.MANUFACTURER).append("\n");
            sb.append("  Android: ").append(Build.VERSION.RELEASE)
              .append(" (API ").append(Build.VERSION.SDK_INT).append(")\n");
            sb.append("  수집시각: ").append(StatsCollector.getCurrentTime()).append("\n");
            sb.append("  실행시간: ").append(StatsCollector.getUptime()).append("\n\n");

            sb.append("━━━ 🔋 배터리 ━━━\n");
            sb.append(StatsCollector.getBatteryInfo(this)).append("\n\n");

            sb.append("━━━ 💻 CPU / 메모리 ━━━\n");
            sb.append("CPU:\n");
            sb.append(StatsCollector.getCpuUsage()).append("\n");
            sb.append("RAM:\n");
            sb.append(StatsCollector.getMemoryInfo(this)).append("\n");
            sb.append("저장공간:\n");
            sb.append(StatsCollector.getStorageInfo()).append("\n\n");

            sb.append("━━━ 🌐 네트워크 ━━━\n");
            sb.append(StatsCollector.getNetworkInfo(this)).append("\n");
            sb.append(StatsCollector.getWifiSignal(this)).append("\n");
            sb.append("블루투스:\n");
            sb.append(StatsCollector.getBluetoothStatus()).append("\n");
            sb.append("화면:\n");
            sb.append(StatsCollector.getScreenStatus(this)).append("\n\n");

            sb.append("━━━ 🔄 실행 중인 서비스 ━━━\n");
            sb.append(StatsCollector.getRunningServices(this)).append("\n\n");

            sb.append("━━━ 📊 최근 사용 앱 (1h) ━━━\n");
            sb.append(StatsCollector.getRecentApps(this)).append("\n\n");

            // UsageStats 권한 상태 안내
            if (!hasUsageStatsPermission()) {
                sb.append("⚠️ 사용량 접근 권한 꺼짐\n");
                sb.append("   → 앱 상단 다이얼로그에서 설정으로 이동 가능\n");
            }

            String info = sb.toString();

            mainHandler.post(() -> {
                tvInfo.setText(info);
                scrollView.scrollTo(0, 0);
                btnRefresh.setEnabled(true);
                btnRefresh.setText("🔄 새로고침");
            });
        });
    }

    private void checkForUpdate() {
        btnUpdateCheck.setEnabled(false);
        btnUpdateCheck.setText("확인 중...");

        executor.execute(() -> {
            String currentVer = BuildConfig.VERSION_NAME;
            UpdateChecker.UpdateInfo info = UpdateChecker.check(currentVer);

            mainHandler.post(() -> {
                btnUpdateCheck.setEnabled(true);
                btnUpdateCheck.setText("🔍 업데이트 확인");

                if (!info.hasUpdate) {
                    if (info.latestVersion == null) {
                        Toast.makeText(this, "업데이트 확인 실패 (네트워크?)", Toast.LENGTH_SHORT).show();
                    } else {
                        Toast.makeText(this, "최신 버전입니다 (" + currentVer + ")", Toast.LENGTH_SHORT).show();
                    }
                    return;
                }

                long sizeKB = info.apkSize / 1024;
                long sizeMB = sizeKB / 1024;
                String sizeStr = sizeMB > 0
                        ? String.format(Locale.US, "%d MB", sizeMB)
                        : String.format(Locale.US, "%d KB", sizeKB);

                new AlertDialog.Builder(this)
                        .setTitle("업데이트 발견 🚀")
                        .setMessage(String.format(Locale.US,
                                "현재: v%s → 최신: v%s\n크기: %s\n\n%s",
                                currentVer, info.latestVersion, sizeStr,
                                info.releaseBody != null ? info.releaseBody : ""))
                        .setPositiveButton("다운로드", (d, w) -> downloadUpdate(info.downloadUrl))
                        .setNegativeButton("취소", null)
                        .show();
            });
        });
    }

    private void downloadUpdate(String url) {
        Toast.makeText(this, "다운로드 시작...", Toast.LENGTH_SHORT).show();
        btnUpdateCheck.setEnabled(false);
        btnUpdateCheck.setText("⬇️ 다운로드 중...");

        executor.execute(() -> {
            UpdateInstaller.downloadAndInstall(url, this, new UpdateInstaller.DownloadCallback() {
                @Override
                public void onProgress(int percent) {
                    mainHandler.post(() ->
                            btnUpdateCheck.setText("⬇️ " + percent + "%"));
                }

                @Override
                public void onComplete(boolean success, String message) {
                    mainHandler.post(() -> {
                        btnUpdateCheck.setEnabled(true);
                        btnUpdateCheck.setText("🔍 업데이트 확인");
                        if (success) {
                            Toast.makeText(MainActivity.this,
                                    "설치를 시작합니다", Toast.LENGTH_LONG).show();
                        } else {
                            Toast.makeText(MainActivity.this,
                                    "실패: " + message, Toast.LENGTH_LONG).show();
                        }
                    });
                }
            });
        });
    }
}

File: app/src/main/java/com/example/phone_stats_collector/MainActivity.kt
package com.example.phone_stats_collector

import android.Manifest
import android.app.AlertDialog
import android.content.Intent
import android.content.pm.PackageManager
import android.os.Build
import android.os.Bundle
import android.os.Handler
import android.os.Looper
import android.provider.Settings
import android.text.method.ScrollingMovementMethod
import android.widget.Button
import android.widget.ScrollView
import android.widget.TextView
import android.widget.Toast
import androidx.appcompat.app.AppCompatActivity
import androidx.core.app.ActivityCompat
import androidx.core.content.ContextCompat
import java.util.Locale
import java.util.concurrent.Executors

@Suppress("DEPRECATION")
class MainActivity : AppCompatActivity() {
    companion object {
        private const val TAG = "MainActivity"
        private const val REQUEST_BLUETOOTH = 1001
        private const val REQUEST_USAGE_STATS = 1002
    }

    private lateinit var tvInfo: TextView
    private lateinit var btnRefresh: Button
    private lateinit var btnUpdateCheck: Button
    private lateinit var scrollView: ScrollView
    private lateinit var btnService: Button

    private val executor = Executors.newSingleThreadExecutor()
    private val mainHandler = Handler(Looper.getMainLooper())

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)

        tvInfo = findViewById(R.id.tv_info)
        btnRefresh = findViewById(R.id.btn_refresh)
        btnUpdateCheck = findViewById(R.id.btn_update)
        scrollView = findViewById(R.id.scroll_view)
        btnService = findViewById(R.id.btn_service)

        tvInfo.movementMethod = ScrollingMovementMethod()
        tvInfo.setTextIsSelectable(true)

        btnRefresh.setOnClickListener { collectStats() }
        btnUpdateCheck.setOnClickListener { checkForUpdate() }
        btnService.setOnClickListener { toggleService() }

        checkUsageStatsPermission()
        collectStats()
        updateServiceButton()
    }

    override fun onResume() {
        super.onResume()
        checkUsageStatsPermission()
        requestBluetoothPermission()
        updateServiceButton()
    }

    private fun toggleService() {
        val intent = Intent(this, StatsApiService::class.java)
        if (StatsApiService.isRunning) {
            stopService(intent)
            Toast.makeText(this, "🛑 API 서버 중지됨", Toast.LENGTH_SHORT).show()
        } else {
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.TIRAMISU) {
                if (ContextCompat.checkSelfPermission(this, Manifest.permission.POST_NOTIFICATIONS)
                    != PackageManager.PERMISSION_GRANTED
                ) {
                    ActivityCompat.requestPermissions(
                        this,
                        arrayOf(Manifest.permission.POST_NOTIFICATIONS),
                        1003
                    )
                    Toast.makeText(this, "알림 권한을 허용해야 서비스가 실행됩니다", Toast.LENGTH_LONG).show()
                    return
                }
            }
            ContextCompat.startForegroundService(this, intent)
            Toast.makeText(this, "🚀 API 서버 시작 (포트 8766)", Toast.LENGTH_SHORT).show()
        }
        updateServiceButton()
    }

    private fun updateServiceButton() {
        btnService.text = if (StatsApiService.isRunning)
            "⏹ 서비스 중지 (8766)" else "▶ 서비스 시작"
    }

    /* ----- 권한 ----- */

    private fun checkUsageStatsPermission() {
        if (Build.VERSION.SDK_INT < Build.VERSION_CODES.LOLLIPOP) return
        if (hasUsageStatsPermission()) return

        Toast.makeText(this, "📊 사용량 접근 권한이 필요합니다", Toast.LENGTH_LONG).show()
        val intent = Intent(Settings.ACTION_USAGE_ACCESS_SETTINGS)
        startActivityForResult(intent, REQUEST_USAGE_STATS)
    }

    private fun requestBluetoothPermission() {
        if (Build.VERSION.SDK_INT < 31) return
        if (ContextCompat.checkSelfPermission(this, Manifest.permission.BLUETOOTH_CONNECT)
            == PackageManager.PERMISSION_GRANTED
        ) return

        if (ActivityCompat.shouldShowRequestPermissionRationale(this, Manifest.permission.BLUETOOTH_CONNECT)) {
            AlertDialog.Builder(this)
                .setTitle("🔵 블루투스 권한")
                .setMessage("블루투스 상태를 표시하려면 권한이 필요합니다")
                .setPositiveButton("허용") { _, _ ->
                    ActivityCompat.requestPermissions(
                        this, arrayOf(Manifest.permission.BLUETOOTH_CONNECT), REQUEST_BLUETOOTH
                    )
                }
                .setNegativeButton("거부", null)
                .show()
        } else {
            ActivityCompat.requestPermissions(
                this, arrayOf(Manifest.permission.BLUETOOTH_CONNECT), REQUEST_BLUETOOTH
            )
        }
    }

    private fun hasUsageStatsPermission(): Boolean {
        if (Build.VERSION.SDK_INT < Build.VERSION_CODES.KITKAT) return true
        return try {
            val aom = getSystemService(APP_OPS_SERVICE) as android.app.AppOpsManager
            aom.checkOpNoThrow(
                android.app.AppOpsManager.OPSTR_GET_USAGE_STATS,
                android.os.Process.myUid(), packageName
            ) == android.app.AppOpsManager.MODE_ALLOWED
        } catch (_: Exception) {
            false
        }
    }

    /* ----- 수집 ----- */

    private fun collectStats() {
        btnRefresh.isEnabled = false
        btnRefresh.text = "수집 중..."

        executor.execute {
            val sb = StringBuilder()

            sb.append("━━━ 📱 기기 정보 ━━━\n")
            sb.append("  기종: ${Build.MODEL}\n")
            sb.append("  제조사: ${Build.MANUFACTURER}\n")
            sb.append("  Android: ${Build.VERSION.RELEASE} (API ${Build.VERSION.SDK_INT})\n")
            sb.append("  수집시각: ${StatsCollector.getCurrentTime()}\n")
            sb.append("  실행시간: ${StatsCollector.getUptime()}\n\n")

            sb.append("━━━ 🔋 배터리 ━━━\n")
            sb.append("${StatsCollector.getBatteryInfo(this)}\n\n")

            sb.append("━━━ 💻 CPU / 메모리 ━━━\n")
            sb.append("CPU:\n${StatsCollector.getCpuUsage()}\n")
            sb.append("RAM:\n${StatsCollector.getMemoryInfo(this)}\n")
            sb.append("저장공간:\n${StatsCollector.getStorageInfo()}\n\n")

            sb.append("━━━ 🌐 네트워크 ━━━\n")
            sb.append("${StatsCollector.getNetworkInfo(this)}\n")
            sb.append("${StatsCollector.getWifiSignal(this)}\n")
            sb.append("블루투스:\n${StatsCollector.getBluetoothStatus()}\n")
            sb.append("화면:\n${StatsCollector.getScreenStatus(this)}\n\n")

            sb.append("━━━ 🔄 실행 중인 서비스 ━━━\n")
            sb.append("${StatsCollector.getRunningServices(this)}\n\n")

            sb.append("━━━ 📊 최근 사용 앱 (1h) ━━━\n")
            sb.append("${StatsCollector.getRecentApps(this)}\n\n")

            sb.append("━━━ 🌐 API 서버 ━━━\n")
            sb.append(if (StatsApiService.isRunning)
                "  실행 중 (포트 8766) ✅\n"
            else
                "  꺼짐 ❌\n")
            sb.append("  Termux: curl http://127.0.0.1:8766/api/stats\n\n")

            if (!hasUsageStatsPermission()) {
                sb.append("⚠️ 사용량 접근 권한 꺼짐\n")
                sb.append("   → 자동으로 설정 페이지 열었음, 활성화 후 돌아오면 반영됨\n")
            }

            val info = sb.toString()

            mainHandler.post {
                tvInfo.text = info
                scrollView.scrollTo(0, 0)
                btnRefresh.isEnabled = true
                btnRefresh.text = "🔄 새로고침"
            }
        }
    }

    /* ----- 업데이트 ----- */

    private fun checkForUpdate() {
        btnUpdateCheck.isEnabled = false
        btnUpdateCheck.text = "확인 중..."

        executor.execute {
            val currentVer = BuildConfig.VERSION_NAME
            val info = UpdateChecker.check(currentVer)

            mainHandler.post {
                btnUpdateCheck.isEnabled = true
                btnUpdateCheck.text = "🔍 업데이트 확인"

                if (!info.hasUpdate) {
                    if (info.latestVersion == null) {
                        Toast.makeText(this@MainActivity, "업데이트 확인 실패 (네트워크?)", Toast.LENGTH_SHORT).show()
                    } else {
                        Toast.makeText(this@MainActivity, "최신 버전입니다 ($currentVer)", Toast.LENGTH_SHORT).show()
                    }
                    return@post
                }

                val sizeKB = info.apkSize / 1024
                val sizeMB = sizeKB / 1024
                val sizeStr = if (sizeMB > 0) "${sizeMB} MB" else "${sizeKB} KB"

                AlertDialog.Builder(this@MainActivity)
                    .setTitle("업데이트 발견 🚀")
                    .setMessage(String.format(Locale.US,
                        "현재: v%s → 최신: v%s\n크기: %s\n\n%s",
                        currentVer, info.latestVersion, sizeStr, info.releaseBody ?: ""))
                    .setPositiveButton("다운로드") { _, _ -> downloadUpdate(info.downloadUrl) }
                    .setNegativeButton("취소", null)
                    .show()
            }
        }
    }

    private fun downloadUpdate(url: String) {
        Toast.makeText(this, "다운로드 시작...", Toast.LENGTH_SHORT).show()
        btnUpdateCheck.isEnabled = false
        btnUpdateCheck.text = "⬇️ 다운로드 중..."

        executor.execute {
            UpdateInstaller.downloadAndInstall(url, this@MainActivity,
                object : UpdateInstaller.DownloadCallback {
                    override fun onProgress(percent: Int) {
                        mainHandler.post {
                            btnUpdateCheck.text = "⬇️ $percent%"
                        }
                    }

                    override fun onComplete(success: Boolean, message: String?) {
                        mainHandler.post {
                            btnUpdateCheck.isEnabled = true
                            btnUpdateCheck.text = "🔍 업데이트 확인"
                            if (success) {
                                Toast.makeText(this@MainActivity, "설치를 시작합니다", Toast.LENGTH_LONG).show()
                            } else {
                                Toast.makeText(this@MainActivity, "실패: $message", Toast.LENGTH_LONG).show()
                            }
                        }
                    }
                })
        }
    }
}